File: app-ios/CEAD/Vistas/Pantallas.swift
import SwiftUI
import CeadKit

// MARK: - Horario

struct PantallaHorario: View {
    let horario: Horario

    var body: some View {
        if horario.curso == nil {
            Vacio(
                titulo: "Todavía no estás en un curso",
                detalle: "Cuando secretaría te asigne a uno, tu horario aparece acá."
            )
        } else if horario.dias.isEmpty {
            Vacio(
                titulo: "Sin horario cargado",
                detalle: "El horario de \(horario.curso?.titulo ?? "tu curso") todavía no fue cargado."
            )
        } else {
            List {
                ForEach(horario.dias) { dia in
                    Section(dia.nombre) {
                        ForEach(Array(dia.franjas.enumerated()), id: \.offset) { _, f in
                            HStack(alignment: .top, spacing: 12) {
                                Text(f.inicio.isEmpty ? "--:--" : f.inicio)
                                    .font(.callout.monospacedDigit())
                                    .foregroundStyle(.secondary)
                                VStack(alignment: .leading, spacing: 2) {
                                    Text(f.materia).font(.body.weight(.medium))
                                    let detalle = [f.docente, f.aula]
                                        .filter { !$0.isEmpty }
                                        .joined(separator: " · ")
                                    if !detalle.isEmpty {
                                        Text(detalle).font(.caption).foregroundStyle(.secondary)
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}

// MARK: - Comunicados

struct PantallaComunicados: View {
    let datos: Comunicados
    let abrir: (Int) -> Void

    var body: some View {
        if datos.comunicados.isEmpty {
            Vacio(
                titulo: "Sin comunicados",
                detalle: "Cuando el colegio publique algo, lo vas a ver acá."
            )
        } else {
            List(datos.comunicados) { c in
                Button { abrir(c.id) } label: {
                    VStack(alignment: .leading, spacing: 4) {
                        if let cat = c.categoria {
                            Text(cat.nombre.uppercased())
                                .font(.caption2.weight(.semibold))
                                .foregroundStyle(Color.ceadMarca)
                        }
                        // Lo no leído pesa más: es la única señal que distingue
                        // lo que falta mirar de lo que ya se miró.
                        Text(c.titulo)
                            .font(.headline)
                            .fontWeight(c.leido ? .regular : .bold)
                        if !c.resumen.isEmpty {
                            Text(c.resumen)
                                .font(.subheadline)
                                .foregroundStyle(.secondary)
                                .lineLimit(3)
                        }
                    }
                }
                .buttonStyle(.plain)
            }
        }
    }
}

struct VistaComunicado: View {
    let comunicado: Comunicado

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 12) {
                Text(comunicado.titulo).font(.title2.bold())
                // El cuerpo viene como HTML del editor de WordPress. Se
                // convierte a texto con formato en vez de mostrar las etiquetas
                // crudas, y sin WebView: no hace falta un navegador entero para
                // leer un comunicado.
                Text(textoDelCuerpo)
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding()
        }
    }

    private var textoDelCuerpo: AttributedString {
        let html = comunicado.contenido ?? comunicado.resumen
        guard let datos = html.data(using: .utf8),
              let attr = try? NSAttributedString(
                  data: datos,
                  options: [
                      .documentType: NSAttributedString.DocumentType.html,
                      .characterEncoding: String.Encoding.utf8.rawValue,
                  ],
                  documentAttributes: nil
              ) else {
            // Si el HTML no se puede interpretar, mejor el texto crudo que una
            // pantalla vacía: el comunicado igual se lee.
            return AttributedString(html)
        }
        return AttributedString(attr)
    }
}

// MARK: - Boletín

struct PantallaBoletin: View {
    let boletin: Boletin

    var body: some View {
        if boletin.materias.isEmpty {
            Vacio(
                titulo: "Sin calificaciones cargadas",
                detalle: "Cuando dirección o tus docentes carguen notas, vas a verlas acá."
            )
        } else {
            List(boletin.materias) { m in
                Section(m.materia) {
                    // Se recorre la lista de períodos del servidor y no las
                    // claves del diccionario: así todas las materias se leen en
                    // la misma columna aunque a alguna le falte un período.
                    ForEach(boletin.periodos, id: \.self) { periodo in
                        HStack {
                            Text(periodo)
                            Spacer()
                            Text(m.notas[periodo]?.texto ?? "—")
                                .fontWeight(.medium)
                                .monospacedDigit()
                        }
                    }
                }
            }
        }
    }
}

// MARK: - Tareas

struct PantallaTareas: View {
    let datos: Tareas

    private var pendientes: [Tarea] { datos.tareas.filter { !$0.hecha } }
    private var hechas: [Tarea] { datos.tareas.filter(\.hecha) }

    var body: some View {
        if datos.tareas.isEmpty {
            Vacio(titulo: "Sin tareas", detalle: "No tenés tareas cargadas en este momento.")
        } else {
            List {
                if !pendientes.isEmpty {
                    Section("Pendientes") {
                        ForEach(pendientes) { FilaTarea(tarea: $0) }
                    }
                }
                if !hechas.isEmpty {
                    Section("Hechas") {
                        ForEach(hechas) { FilaTarea(tarea: $0) }
                    }
                }
            }
        }
    }
}

private struct FilaTarea: View {
    let tarea: Tarea

    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            Text(tarea.titulo)
                .foregroundStyle(tarea.hecha ? .secondary : .primary)
            let pie = [
                tarea.vence.flatMap { $0.isEmpty ? nil : "Vence \($0)" },
                tarea.prioridad == "alta" ? "Prioridad alta" : nil,
            ].compactMap { $0 }.joined(separator: " · ")
            if !pie.isEmpty {
                Text(pie).font(.caption).foregroundStyle(.secondary)
            }
        }
    }
}

File: app-android/app/src/main/java/net/caaguazu/cead/panel/ui/Pantallas.kt
package net.caaguazu.cead.panel.ui

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.Card
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import net.caaguazu.cead.panel.data.Boletin
import net.caaguazu.cead.panel.data.Comunicados
import net.caaguazu.cead.panel.data.Horario
import net.caaguazu.cead.panel.data.Tareas

/* ------------------------------------------------------------------ horario */

@Composable
fun PantallaHorario(horario: Horario) {
    if (horario.curso == null) {
        Vacio(
            "Todavía no estás en un curso",
            "Cuando secretaría te asigne a uno, tu horario aparece acá.",
        )
        return
    }
    if (horario.dias.isEmpty()) {
        Vacio(
            "Sin horario cargado",
            "El horario de ${horario.curso.titulo} todavía no fue cargado.",
        )
        return
    }

    LazyColumn(
        contentPadding = androidx.compose.foundation.layout.PaddingValues(16.dp),
        verticalArrangement = Arrangement.spacedBy(16.dp),
    ) {
        item {
            Text(horario.curso.titulo, style = MaterialTheme.typography.titleLarge)
        }
        items(horario.dias) { dia ->
            Card(modifier = Modifier.fillMaxWidth()) {
                Column(Modifier.padding(16.dp)) {
                    Text(nombreDelDia(dia.dia), style = MaterialTheme.typography.titleMedium)
                    Spacer(Modifier.height(8.dp))
                    dia.franjas.forEachIndexed { i, f ->
                        if (i > 0) HorizontalDivider(Modifier.padding(vertical = 8.dp))
                        Row(verticalAlignment = Alignment.CenterVertically) {
                            Text(
                                f.inicio.ifBlank { "--:--" },
                                style = MaterialTheme.typography.labelLarge,
                                modifier = Modifier.padding(end = 12.dp),
                            )
                            Column {
                                Text(f.materia, fontWeight = FontWeight.Medium)
                                val detalle = listOfNotNull(
                                    f.docente.takeIf { it.isNotBlank() },
                                    f.aula.takeIf { it.isNotBlank() },
                                ).joinToString(" · ")
                                if (detalle.isNotBlank()) {
                                    Text(
                                        detalle,
                                        style = MaterialTheme.typography.bodySmall,
                                        color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.7f),
                                    )
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}

/* -------------------------------------------------------------- comunicados */

@Composable
fun PantallaComunicados(datos: Comunicados, onAbrir: (Int) -> Unit) {
    if (datos.comunicados.isEmpty()) {
        Vacio("Sin comunicados", "Cuando el colegio publique algo, lo vas a ver acá.")
        return
    }

    LazyColumn(
        contentPadding = androidx.compose.foundation.layout.PaddingValues(16.dp),
        verticalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        items(datos.comunicados) { c ->
            Card(
                modifier = Modifier.fillMaxWidth().clickable { onAbrir(c.id) },
            ) {
                Column(Modifier.padding(16.dp)) {
                    if (c.categoria != null) {
                        Text(
                            c.categoria.nombre.uppercase(),
                            style = MaterialTheme.typography.labelSmall,
                            color = MaterialTheme.colorScheme.primary,
                        )
                    }
                    Text(
                        c.titulo,
                        style = MaterialTheme.typography.titleMedium,
                        // Lo no leído pesa más: es la única señal que distingue
                        // lo que falta mirar de lo que ya se miró.
                        fontWeight = if (c.leido) FontWeight.Normal else FontWeight.Bold,
                    )
                    if (c.resumen.isNotBlank()) {
                        Spacer(Modifier.height(4.dp))
                        Text(
                            c.resumen,
                            style = MaterialTheme.typography.bodyMedium,
                            maxLines = 3,
                            color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.8f),
                        )
                    }
                }
            }
        }
    }
}

/* ------------------------------------------------------------------ boletín */

@Composable
fun PantallaBoletin(boletin: Boletin) {
    if (boletin.materias.isEmpty()) {
        Vacio(
            "Sin calificaciones cargadas",
            "Cuando dirección o tus docentes carguen notas, vas a verlas acá.",
        )
        return
    }

    LazyColumn(
        contentPadding = androidx.compose.foundation.layout.PaddingValues(16.dp),
        verticalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        items(boletin.materias) { m ->
            Card(modifier = Modifier.fillMaxWidth()) {
                Column(Modifier.padding(16.dp)) {
                    Text(m.materia, style = MaterialTheme.typography.titleMedium)
                    Spacer(Modifier.height(8.dp))
                    // Se recorre la lista de períodos del servidor y no las
                    // claves del mapa: así todas las materias se leen en la
                    // misma columna aunque a alguna le falte un período.
                    boletin.periodos.forEach { periodo ->
                        val n = m.notas[periodo]
                        Row(
                            modifier = Modifier.fillMaxWidth().padding(vertical = 2.dp),
                            horizontalArrangement = Arrangement.SpaceBetween,
                        ) {
                            Text(periodo, style = MaterialTheme.typography.bodyMedium)
                            Text(
                                n?.let { it.letra.ifBlank { it.nota?.let { v -> formatearNota(v) } ?: "—" } } ?: "—",
                                style = MaterialTheme.typography.bodyMedium,
                                fontWeight = FontWeight.Medium,
                            )
                        }
                    }
                }
            }
        }
    }
}

/** Sin decimales cuando no hacen falta: «4» se lee mejor que «4.0». */
private fun formatearNota(v: Double): String =
    if (v == v.toLong().toDouble()) v.toLong().toString() else v.toString()

/* ------------------------------------------------------------------- tareas */

@Composable
fun PantallaTareas(datos: Tareas) {
    if (datos.tareas.isEmpty()) {
        Vacio("Sin tareas", "No tenés tareas cargadas en este momento.")
        return
    }

    val pendientes = datos.tareas.filterNot { it.hecha }
    val hechas = datos.tareas.filter { it.hecha }

    LazyColumn(
        contentPadding = androidx.compose.foundation.layout.PaddingValues(16.dp),
        verticalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        if (pendientes.isNotEmpty()) {
            item { Text("Pendientes", style = MaterialTheme.typography.titleMedium) }
            items(pendientes) { t -> FilaTarea(t.titulo, t.vence, t.prioridad, false) }
        }
        if (hechas.isNotEmpty()) {
            item {
                Spacer(Modifier.height(8.dp))
                Text("Hechas", style = MaterialTheme.typography.titleMedium)
            }
            items(hechas) { t -> FilaTarea(t.titulo, t.vence, t.prioridad, true) }
        }
    }
}

@Composable
private fun FilaTarea(titulo: String, vence: String?, prioridad: String, hecha: Boolean) {
    Card(modifier = Modifier.fillMaxWidth()) {
        Column(Modifier.padding(16.dp)) {
            Text(
                titulo,
                style = MaterialTheme.typography.bodyLarge,
                color = if (hecha) {
                    MaterialTheme.colorScheme.onSurface.copy(alpha = 0.5f)
                } else {
                    MaterialTheme.colorScheme.onSurface
                },
            )
            val pie = listOfNotNull(
                vence?.takeIf { it.isNotBlank() }?.let { "Vence $it" },
                prioridad.takeIf { it == "alta" }?.let { "Prioridad alta" },
            ).joinToString(" · ")
            if (pie.isNotBlank()) {
                Spacer(Modifier.height(4.dp))
                Text(
                    pie,
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.7f),
                )
            }
        }
    }
}
